Consul 0.6.4 crash

If you have a question, please direct it to the
[consul mailing list](https://www.consul.io/community.html) if it hasn't been
addressed in either the [FAQ](https://www.consul.io/docs/faq.html) or in one
of the [Consul Guides](https://www.consul.io/docs/guides/index.html).

When filing a bug, please include the following:

### `consul version` for both Client and Server
Client: 0.6.4
Server: 0.6.4

### `consul info` for both Client and Server
agent:
        check_monitors = 0
        check_ttls = 0
        checks = 22
        services = 24
build:
        prerelease = 
        revision = 26a0ef8c
        version = 0.6.4
consul:
        bootstrap = false
        known_datacenters = 1
        leader = false
        server = true
raft:
        applied_index = 999
        commit_index = 999
        fsm_pending = 0
        last_contact = 20.884238ms
        last_log_index = 999
        last_log_term = 3
        last_snapshot_index = 0
        last_snapshot_term = 0
        num_peers = 12
        state = Follower
        term = 3
runtime:
        arch = amd64
        cpu_count = 4
        goroutines = 86
        max_procs = 2
        os = linux
        version = go1.6
serf_lan:
        encrypted = true
        event_queue = 0
        event_time = 3
        failed = 0
        intent_queue = 0
        left = 0
        member_time = 16
        members = 13
        query_queue = 0
        query_time = 1
serf_wan:
        encrypted = true
        event_queue = 0
        event_time = 1
        failed = 0
        intent_queue = 0
        left = 0
        member_time = 1
        members = 1
        query_queue = 0
        query_time = 1

### Operating system and Environment details
Amazon Linux update to date with all update

### Description of the Issue (and unexpected/desired result)
Crash of consul

### Reproduction steps
Not able to reproducre

### Log Fragments or Link to [gist](https://gist.github.com/)
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x68 pc=0x4f00e0]

goroutine 20 [running]:
panic(0xcb9ec0, 0xc82000e0c0)
        /goroot/src/runtime/panic.go:464 +0x3e6
github.com/hashicorp/consul/command/agent.(*localState).syncCheck(0xc820001cb0, 0xc822dd35c0, 0x5c, 0x0, 0x0)
        /gopath/src/github.com/hashicorp/consul/command/agent/local.go:606 +0x90
github.com/hashicorp/consul/command/agent.(*localState).syncChanges(0xc820001cb0, 0x0, 0x0)
        /gopath/src/github.com/hashicorp/consul/command/agent/local.go:488 +0x446
github.com/hashicorp/consul/command/agent.(*localState).antiEntropy(0xc820001cb0, 0xc82000acc0)
        /gopath/src/github.com/hashicorp/consul/command/agent/local.go:339 +0x208
created by github.com/hashicorp/consul/command/agent.(*Agent).StartSync
        /gopath/src/github.com/hashicorp/consul/command/agent/agent.go:580 +0x4f

We have the same issue, consul crash ~300 times a day !!!! , on all cluster nodes ( we have 3)
the error message:
2016/06/30 03:44:22 [INFO] snapshot: Creating new snapshot at /opt/consul/data/raft/snapshots/16018-198934171-1467272662110.tmp
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x28 pc=0x9ee7f9]

Please Advise
Thanks
Riki

--- a/command/agent/local.go
+++ b/command/agent/local.go
@@ -26,8 +26,7 @@ const (
 // syncStatus is used to represent the difference between
 // the local and remote state, and if action needs to be taken
 type syncStatus struct {
-	remoteDelete bool // Should this be deleted from the server
-	inSync       bool // Is this in sync with the server
+	inSync bool // Is this in sync with the server
 }
 
 // localState is used to represent the node's services,
@@ -177,7 +176,7 @@ func (l *localState) RemoveService(serviceID string) {
 
 	delete(l.services, serviceID)
 	delete(l.serviceTokens, serviceID)
-	l.serviceStatus[serviceID] = syncStatus{remoteDelete: true}
+	l.serviceStatus[serviceID] = syncStatus{inSync: false}
 	l.changeMade()
 }
 
@@ -237,7 +236,7 @@ func (l *localState) RemoveCheck(checkID types.CheckID) {
 	delete(l.checks, checkID)
 	delete(l.checkTokens, checkID)
 	delete(l.checkCriticalTime, checkID)
-	l.checkStatus[checkID] = syncStatus{remoteDelete: true}
+	l.checkStatus[checkID] = syncStatus{inSync: false}
 	l.changeMade()
 }
 
@@ -434,7 +433,7 @@ func (l *localState) setSyncState() error {
 		// If we don't have the service locally, deregister it
 		existing, ok := l.services[id]
 		if !ok {
-			l.serviceStatus[id] = syncStatus{remoteDelete: true}
+			l.serviceStatus[id] = syncStatus{inSync: false}
 			continue
 		}
 
@@ -472,7 +471,7 @@ func (l *localState) setSyncState() error {
 			if id == consul.SerfCheckID {
 				continue
 			}
-			l.checkStatus[id] = syncStatus{remoteDelete: true}
+			l.checkStatus[id] = syncStatus{inSync: false}
 			continue
 		}
 
@@ -521,7 +520,7 @@ func (l *localState) syncChanges() error {
 
 	// Sync the services
 	for id, status := range l.serviceStatus {
-		if status.remoteDelete {
+		if _, ok := l.services[id]; !ok {
 			if err := l.deleteService(id); err != nil {
 				return err
 			}
@@ -536,7 +535,7 @@ func (l *localState) syncChanges() error {
 
 	// Sync the checks
 	for id, status := range l.checkStatus {
-		if status.remoteDelete {
+		if _, ok := l.checks[id]; !ok {
 			if err := l.deleteCheck(id); err != nil {
 				return err
 			}

--- a/command/agent/local_test.go
+++ b/command/agent/local_test.go
@@ -175,6 +175,63 @@ func TestAgentAntiEntropy_Services(t *testing.T) {
 			t.Fatalf("should be in sync: %v %v", name, status)
 		}
 	}
+
+	// Remove one of the services
+	agent.state.RemoveService("api")
+
+	// Trigger anti-entropy run and wait
+	agent.StartSync()
+	time.Sleep(200 * time.Millisecond)
+
+	// Verify that we are in sync
+	if err := agent.RPC("Catalog.NodeServices", &req, &services); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// We should have 5 services (consul included)
+	if len(services.NodeServices.Services) != 5 {
+		t.Fatalf("bad: %v", services.NodeServices.Services)
+	}
+
+	// All the services should match
+	for id, serv := range services.NodeServices.Services {
+		serv.CreateIndex, serv.ModifyIndex = 0, 0
+		switch id {
+		case "mysql":
+			if !reflect.DeepEqual(serv, srv1) {
+				t.Fatalf("bad: %v %v", serv, srv1)
+			}
+		case "redis":
+			if !reflect.DeepEqual(serv, srv2) {
+				t.Fatalf("bad: %#v %#v", serv, srv2)
+			}
+		case "web":
+			if !reflect.DeepEqual(serv, srv3) {
+				t.Fatalf("bad: %v %v", serv, srv3)
+			}
+		case "cache":
+			if !reflect.DeepEqual(serv, srv6) {
+				t.Fatalf("bad: %v %v", serv, srv6)
+			}
+		case "consul":
+			// ignore
+		default:
+			t.Fatalf("unexpected service: %v", id)
+		}
+	}
+
+	// Check the local state
+	if len(agent.state.services) != 5 {
+		t.Fatalf("bad: %v", agent.state.services)
+	}
+	if len(agent.state.serviceStatus) != 5 {
+		t.Fatalf("bad: %v", agent.state.serviceStatus)
+	}
+	for name, status := range agent.state.serviceStatus {
+		if !status.inSync {
+			t.Fatalf("should be in sync: %v %v", name, status)
+		}
+	}
 }
 
 func TestAgentAntiEntropy_EnableTagOverride(t *testing.T) {
@@ -649,6 +706,59 @@ func TestAgentAntiEntropy_Checks(t *testing.T) {
 		addrs := services.NodeServices.Node.TaggedAddresses
 		if len(addrs) == 0 || !reflect.DeepEqual(addrs, conf.TaggedAddresses) {
 			t.Fatalf("bad: %v", addrs)
+		}
+	}
+
+	// Remove one of the checks
+	agent.state.RemoveCheck("redis")
+
+	// Trigger anti-entropy run and wait
+	agent.StartSync()
+	time.Sleep(200 * time.Millisecond)
+
+	// Verify that we are in sync
+	if err := agent.RPC("Health.NodeChecks", &req, &checks); err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	// We should have 5 checks (serf included)
+	if len(checks.HealthChecks) != 4 {
+		t.Fatalf("bad: %v", checks)
+	}
+
+	// All the checks should match
+	for _, chk := range checks.HealthChecks {
+		chk.CreateIndex, chk.ModifyIndex = 0, 0
+		switch chk.CheckID {
+		case "mysql":
+			if !reflect.DeepEqual(chk, chk1) {
+				t.Fatalf("bad: %v %v", chk, chk1)
+			}
+		case "web":
+			if !reflect.DeepEqual(chk, chk3) {
+				t.Fatalf("bad: %v %v", chk, chk3)
+			}
+		case "cache":
+			if !reflect.DeepEqual(chk, chk5) {
+				t.Fatalf("bad: %v %v", chk, chk5)
+			}
+		case "serfHealth":
+			// ignore
+		default:
+			t.Fatalf("unexpected check: %v", chk)
+		}
+	}
+
+	// Check the local state
+	if len(agent.state.checks) != 3 {
+		t.Fatalf("bad: %v", agent.state.checks)
+	}
+	if len(agent.state.checkStatus) != 3 {
+		t.Fatalf("bad: %v", agent.state.checkStatus)
+	}
+	for name, status := range agent.state.checkStatus {
+		if !status.inSync {
+			t.Fatalf("should be in sync: %v %v", name, status)
 		}
 	}
 }

